templates/company-board/contracts/CompanyBoardTemplate.sol

```diff
@@ -42,207 +42,173 @@ contract CompanyBoardTemplate is BaseTemplate {
+    function _popDaoCache() internal returns (Kernel dao, Voting shareVoting, Voting boardVoting) {
         Cache storage c = cache[msg.sender];
-        require(c.dao != address(0), ERROR_MISSING_CACHE);
-
-        Kernel dao = Kernel(c.dao);
-        delete c.dao;
-        return dao;
-    }
-
-    function _popBoardTokenCache() internal returns (MiniMeToken) {
-        Cache storage c = cache[msg.sender];
-        require(c.boardToken != address(0), ERROR_MISSING_CACHE);
-
-        MiniMeToken boardToken = MiniMeToken(c.boardToken);
-        delete c.boardToken;
-        return boardToken;
-    }
-
-    function _popShareTokenCache() internal returns (MiniMeToken) {
-        Cache storage c = cache[msg.sender];
-        require(c.shareToken != address(0), ERROR_MISSING_CACHE);
-
-        MiniMeToken shareToken = MiniMeToken(c.shareToken);
-        delete c.shareToken;
-        return shareToken;
-    }
-
-    function _popBoardApps() internal returns (Voting boardVoting, TokenManager boardTokenManager) {
-        Cache storage c = cache[msg.sender];
-        require(c.boardVoting != address(0) && c.boardTokenManager != address(0), ERROR_MISSING_CACHE);
+        require(c.dao != address(0) && c.shareVoting != address(0) && c.boardVoting != address(0), ERROR_MISSING_CACHE);
 
+        dao = Kernel(c.dao);
+        shareVoting = Voting(c.shareVoting);
         boardVoting = Voting(c.boardVoting);
-        boardTokenManager = TokenManager(c.boardTokenManager);
+        delete c.dao;
+        delete c.shareVoting;
         delete c.boardVoting;
-        delete c.boardTokenManager;
     }
 
-    function _fetchDaoCache() internal view returns (Kernel) {
+    function _popTokenCaches() internal returns (MiniMeToken shareToken, MiniMeToken boardToken) {
         Cache storage c = cache[msg.sender];
-        require(c.dao != address(0), ERROR_MISSING_CACHE);
-        return Kernel(c.dao);
-    }
+        require(c.shareToken != address(0) && c.boardToken != address(0), ERROR_MISSING_CACHE);
 
-    function _ensureCompanySettings(address[] _holders, uint256[] _stakes, uint64[3] _votingSettings, uint256[4] _payrollSettings) private pure {
-        _ensureCompanySettings(_holders, _stakes, _votingSettings);
-        require(_payrollSettings.length == 4, ERROR_BAD_PAYROLL_SETTINGS);
+        shareToken = MiniMeToken(c.shareToken);
+        boardToken = MiniMeToken(c.boardToken);
+        delete c.shareToken;
+        delete c.boardToken;
     }
 
-    function _ensureCompanySettings(address[] _holders, uint256[] _stakes, uint64[3] _votingSettings) private pure {
-        require(_holders.length > 0, ERROR_MISSING_SHARE_MEMBERS);
-        require(_holders.length == _stakes.length, ERROR_BAD_HOLDERS_STAKES_LEN);
-        require(_votingSettings.length == 3, ERROR_BAD_VOTE_SETTINGS);
+    function _ensureFinalizationSettings(address[] memory _shareHolders, uint256[] memory _shareStakes, address[] memory _boardMembers) internal {
```

Comment:
I've added `memory` to the internal functions passing along arrays; what do you think about doing so for the rest of the templates that use this as well?

Reply:
It could help when we migrate to Solidity 0.5.x. If we do it, we should also add it for any parameters of type struct, mapping and string.